src/main.ts

```diff
@@ -130,41 +121,41 @@ async function processTransfersERC721(ctx: Context, transfersData: TransferEvent
 }
 
 
-export async function getOrCreateNfts(ctx: Context, cache: Cache, nftsData: Map<string, NftData>){
+export async function getOrCreateNfts(ctx: Context, latestBlockNumber: number, cache: Cache, nftsData: Map<string, TransferEvent>){
     for (let batch of splitIntoBatches([...nftsData.keys()])){
         (await ctx.store.findBy(NftEntity, { id: In(batch) })).map((entity) => {
             cache.Nfts.set(entity.id, entity)
             nftsData.delete(entity.id)
         });
     }
 
-    const collectionsData: Map<string, CollectionData> = new Map()
+    const collectionsData: Map<string, TransferEvent> = new Map()
     for(const nftData of nftsData.values()){
         collectionsData.set(
             getCollectionEntityId(nftData.contractAddress, nftData.blockchain),
-            {
-                contractAddress: nftData.contractAddress,
-                blockchain: nftData.blockchain,
-                contractType: nftData.contractType
-            }
+            nftData
         )
     }
-    await getOrCreateNftCollections(ctx, cache, collectionsData)
-
-    for(const[nftId, nftData] of nftsData){
-       const collectionId = getCollectionEntityId(nftData.contractAddress, nftData.blockchain)
-       const nftCollenction = cache.NftCollections.get(collectionId)
-       let nftEntity = new NftEntity({
-            id: nftId,
-            tokenId: nftData.tokenId,
-            nftCollection: nftCollenction
-       });
-       cache.Nfts.set(nftId, nftEntity);
+    await getOrCreateNftCollections(ctx, latestBlockNumber, cache, collectionsData)
+    for (let batch of splitIntoBatches([...nftsData.entries()], 25)){
+        Promise.all(batch.map(async ([nftId, nftData]) => {
+            const collectionId = getCollectionEntityId(nftData.contractAddress, nftData.blockchain)
+            const nftCollenction = cache.NftCollections.get(collectionId)
+            let nftEntity = new NftEntity({
+                 id: nftId,
+                 tokenId: nftData.tokenIds[0],
+                 nftCollection: nftCollenction
+            });
+            fillTokenMetadata(ctx, latestBlockNumber, nftEntity);
+            cache.Nfts.set(nftId, nftEntity);
+        }))
+        await new Promise(f => setTimeout(f, 10000));
+
     }
 }
 
 
-export async function getOrCreateNftCollections(ctx: Context, cache: Cache, collectionsData: Map<string, CollectionData>){
+export async function getOrCreateNftCollections(ctx: Context, latestBlockNumber: number, cache: Cache, collectionsData: Map<string, TransferEvent>){
     for (let batch of splitIntoBatches([...collectionsData.keys()])){
```

Comment:
Again

Reply:
spread is used to iterate the iterator and convert to array